Command to generate pipeline container config files

mirpedrol opened the pull request:
close #3953 

Done:

- New function to run `nextflow inspect` and obtain docker config file
- New function to (re)generate all container config files
- Container config files are regenerated with `nf-core modules install|update|remove`

TODO:

- [x] Add tests
- [x] Test a lot
- [ ] What happens if the module is not found? 
    - Modules form multiple repos
    - Modules included as an alias
- [ ] ~Is there a case where two modules have the same name but different container? how to solve this?~ We couldn't find an existing case of this
- [ ] Could the code be used to reduced duplication with download command?

mirpedrol committed:
initial functions to generate pipeline container config files

mirpedrol committed:
generate all config files

mirpedrol committed:
regenerate container config files when modules install, remove, update

mirpedrol committed:
pass repo_path

mirpedrol committed:
add tests

mirpedrol committed:
less lines

mashehu commented on nf_core/pipelines/containers_utils.py:
```diff
@@ -0,0 +1,132 @@
+
+            for p_name, p_keys in platforms.items():
+                try:
+                    containers[p_name][module_name] = meta["containers"][p_keys[0]][p_keys[1]][p_keys[2]]
```
```suggestion
            for p_name, (runtime, arch, protocol) in platforms.items():
			    try:
			        containers[p_name][module_name] = meta["containers"][runtime][arch][protocol]
```
makes it a bit more readable

mirpedrol committed:
always update docker amd64

mirpedrol committed:
Update nf_core/pipelines/containers_utils.py

Co-authored-by: Matthias Hörtenhuber <mashehu@users.noreply.github.com>

mirpedrol committed:
spaces

mashehu commented on nf_core/pipelines/containers_utils.py:
```diff
@@ -0,0 +1,133 @@
+
+    def __init__(
+        self,
+        workflow_directory: str | Path = ".",
```
let's make this simpler

mashehu commented on nf_core/pipelines/containers_utils.py:
```diff
@@ -0,0 +1,133 @@
+        workflow_directory: str | Path = ".",
+        org: str = "nf-core",
+    ):
+        self.workflow_directory = Path(workflow_directory)
```
```suggestion
        self.workflow_directory = workflow_directory
```

mashehu commented on nf_core/components/components_utils.py:
```diff
@@ -279,3 +280,10 @@ def _iterate_input_output(type) -> DictWithStrAndTuple:
     return None
+
+
+def try_generate_container_configs(directory: str | Path, path: str):
```
```suggestion
def try_generate_container_configs(directory: Path, path: Path):
```
let's make this simpler

mashehu commented on nf_core/pipelines/containers_utils.py:
```diff
@@ -0,0 +1,133 @@
+            cmd_params = f"inspect -format config {self.workflow_directory}"
+            cmd_out = run_cmd(executable, cmd_params)
+            if cmd_out is None:
+                raise UserWarning("Failed to run `nextflow inspect`. Please check your Nextflow installation.")
```
I have a feeling we have this multiple times in our code base. should this be a utils function?

mirpedrol replied:
`run_cmd` is the utils function, we raise a warning depending on where we use this function, in some cases we might not want to stop execution. I would say it's fine as it is now

mashehu replied:
I was more thinking about https://github.com/nf-core/tools/pull/3954/changes#diff-8bdd161b806e417fa6f155d1babd55464cbb9cea5f3449392fb23e518b7ca417R363-R369

mirpedrol committed:
Apply suggestions from code review

Co-authored-by: Matthias Hörtenhuber <mashehu@users.noreply.github.com>

mashehu committed:
fix types

mirpedrol committed:
inspect in json format and fix tests

mirpedrol committed:
simplify functions, thanks @mashehu

mirpedrol committed:
Update tests/pipelines/test_container_configs.py

mirpedrol committed:
remove too many warnings and add debug messages and one info summary instead

mashehu commented on nf_core/components/components_utils.py:
```diff
@@ -279,3 +280,10 @@ def _iterate_input_output(type) -> DictWithStrAndTuple:
     return None
+
+
+def try_generate_container_configs(directory: Path, path: Path):
```
```suggestion
def generate_container_configs(directory: Path, path: Path):
```

mashehu commented on nf_core/pipeline-template/workflows/pipeline.nf:
```diff
@@ -43,14 +43,13 @@ workflow {{ short_name|upper }} {
     )
-    {% if multiqc %}ch_multiqc_files = ch_multiqc_files.mix(FASTQC.out.zip.collect{it[1]}){% endif %}
-    ch_versions = ch_versions.mix(FASTQC.out.versions.first())
+    {% if multiqc %}ch_multiqc_files = ch_multiqc_files.mix(FASTQC.out.zip.collect{it -> it[1]}){% endif %}
```
```suggestion
    {% if multiqc %}ch_multiqc_files = ch_multiqc_files.mix(FASTQC.out.zip.collect{file -> file[1]}){% endif %}
```

mashehu committed:
update arch names to make it consistent with modules container command

mashehu committed:
parse the include statements to get modules paths from process names

mashehu committed:
sort generated config files

mashehu committed:
drop `linux_` prefix of arch in config file names

mashehu committed:
add container configs to linting

# Conflicts:
#	nf_core/pipelines/containers_utils.py

mashehu committed:
Apply suggestions from code review

Co-authored-by: Matthias Hörtenhuber <mashehu@users.noreply.github.com>

# Conflicts:
#	nf_core/pipeline-template/workflows/pipeline.nf

mashehu committed:
remove unused argument

mashehu committed:
remove org from containerconfigs

mashehu committed:
update template modules to latest version

mashehu committed:
fix container config test

mashehu committed:
bump nextflow version to 25.10.4

mashehu committed:
fix type warning

mashehu committed:
fix ruff linting errors

mashehu committed:
update modules.json in tempalte

mashehu committed:
fix todo hint

mashehu committed:
add container config creation step to template rendering

mashehu committed:
fix path to mock

mashehu committed:
fix tests by actually removing file that is now generated with pipelines create

mashehu committed:
more test fixes

mashehu committed:
move try_generate_container_configs to container_utils

mashehu committed:
use same PLATFORMS constant in containers_utils and tests

mashehu committed:
also regenerate container configs for patch commands

mashehu committed:
clean up container configs before writing new ones, to keep state in sync

mashehu committed:
manually add newly installed module to container config generator

mashehu committed:
update fastqc and multiqc to latest version

(cherry picked from commit 3e0d9ad69ea762294ea6a6012d4135b44ea937a2)

# Conflicts:
#	nf_core/pipeline-template/modules.json
#	nf_core/pipeline-template/modules/nf-core/fastqc/main.nf
#	nf_core/pipeline-template/modules/nf-core/multiqc/meta.yml
#	nf_core/pipeline-template/modules/nf-core/multiqc/tests/main.nf.test.snap
